pandas/plotting/_matplotlib/hist.py

```diff
@@ -49,14 +49,25 @@ def _plot(
         stacking_id=None,
         **kwds,
     ):
+
+        y = y[~isna(y)]
+
+        # if bins is a string (e.g. "auto"), get actual bins
+        # so that it can be used in creating stackers below
+        _, bins = np.histogram(
+            y,
+            bins=bins,
+            range=kwds.get("range", None),
+            weights=kwds.get("weights", None),
+        )
+
         if column_num == 0:
             cls._initialize_stacker(ax, stacking_id, len(bins) - 1)
-        y = y[~isna(y)]
 
         base = np.zeros(len(bins) - 1)
         bottom = bottom + cls._get_stacked_values(ax, stacking_id, base, kwds["label"])
         # ignore style
-        n, bins, patches = ax.hist(y, bins=bins, bottom=bottom, **kwds)
+        n, _, patches = ax.hist(y, bins=bins, bottom=bottom, **kwds)
```

Comment:
Is there a particular reason that we have a separate call to `np.histogram` above to get the bins? Is it not possible to return the appropriate values from the `ax.hist` call here?

Reply:
`cls._initialize_stacker` needs the number of bins. `ax.hist` can't provide that yet, since it needs `bottom`, which comes from `cls._get_stacked_values` and that can't be called before `cls._initialize_stacker`. That's my understanding.